src/frontend/src/flows/verifiableCredentials/index.ts

```diff
@@ -0,0 +1,415 @@
+import {
+  CredentialSpec,
+  IssuedCredentialData,
+} from "$generated/vc_issuer_types";
+import { handleLogin } from "$src/components/authenticateBox";
+import { withLoader } from "$src/components/loader";
+import { showMessage } from "$src/components/message";
+import { showSpinner } from "$src/components/spinner";
+import { fetchDelegation } from "$src/flows/authorize/fetchDelegation";
+import { AuthenticatedConnection, Connection } from "$src/utils/iiConnection";
+import {
+  Delegation,
+  DelegationChain,
+  DelegationIdentity,
+  ECDSAKeyIdentity,
+} from "@dfinity/identity";
+import { isNullish } from "@dfinity/utils";
+import { base64url } from "jose";
+import { abortedCredentials } from "./abortedCredentials";
+import { allowCredentials } from "./allowCredentials";
+import { VcVerifiablePresentation, vcProtocol } from "./postMessageInterface";
+import { VcIssuer } from "./vcIssuer";
+
+// The "verifiable credentials" approval flow
+export const vcFlow = async ({
+  connection,
+}: {
+  connection: Connection;
+}): Promise<never> => {
+  const result = await vcProtocol({
+    /* Show some spinners while we wait for more data or for an action to complete */
+    onProgress: (x) => {
+      if (x === "waiting") {
+        return showSpinner({
+          message: "Waiting for verification data",
+        });
+      }
+
+      if (x === "verifying") {
+        return showSpinner({
+          message: "Checking verification data",
+        });
+      }
+      x satisfies never;
+    },
+
+    /* How the credentials are actually verified */
+    verifyCredentials: (args) => verifyCredentials({ connection, ...args }),
+  });
+
+  if (result === "orphan") {
+    await showMessage({
+      message: "No credentials data provided! Wrong URL?",
+    });
+  }
+
+  return await new Promise((_) => {
+    /* halt forever */
+  });
+};
+
+type VerifyCredentials = Parameters<typeof vcProtocol>[0]["verifyCredentials"];
+type VerifyCredentialsArgs = Parameters<VerifyCredentials>[0];
+
+const verifyCredentials = async ({
+  connection,
+  request: {
+    credentialSubject: givenP_RP,
+    issuer: { origin: issuerOrigin, canisterId: issuerCanisterId_ },
+    credentialSpec,
+  },
+  rpOrigin,
+}: { connection: Connection } & VerifyCredentialsArgs) => {
+  // Use the issuer canister ID provided by the RP, or look it up
+  // from the issuer origin if no canister ID was provided.
+  // NOTE: we do _not_ check the origin's canister ID if a canister ID was provided explicitly.
+  let issuerCanisterId = issuerCanisterId_?.toText();
+
+  if (isNullish(issuerCanisterId)) {
+    const lookedUp = await withLoader(() =>
+      lookupCanister({ origin: issuerOrigin })
+    );
+    if (lookedUp === "not_found") {
+      return abortedCredentials({ reason: "no_canister_id" });
+    }
+
+    issuerCanisterId = lookedUp.ok;
+  }
+
+  const vcIssuer = new VcIssuer(issuerCanisterId);
+  // XXX: We don't check that the language matches the user's language. We need
```

Comment:
Probably, it does not make sense to support more languages, if the rest of the UI doesn't support it either...